workspaces/ui-v2/src/optic-components/PromptNavigateAway.tsx

```diff
@@ -0,0 +1,25 @@
+import * as React from 'react';
+import { FC, useEffect } from 'react';
+import { Prompt } from 'react-router';
+
+type PromptNavigateAwayProps = {
+  shouldPrompt: boolean;
+};
+
+export const PromptNavigateAway: FC<PromptNavigateAwayProps> = ({
+  shouldPrompt,
+}) => {
+  useEffect(() => {
+    window.onbeforeunload = shouldPrompt ? () => true : null;
```

Comment:
my only nit is that this is destructive to any other handler

Reply:
Hmm great catch - unfortunately I don't have a great solution for this, there's only a single onbeforeunload handler and htere's no guarantee that there'll be any ordering of cleanups (i.e. if I have two promptNavigateAways - they could remove each other in no deterministic order)

I think this is not a likely scenario since onbeforeunload is really only used for prompting - and we'd hopefully only have 1 prompt mounted at a time, but I'll add a comment here noting that this is destructive

Reply:
i'm pretty sure i've seen a library that provides a component and context supporting a stack of these types of handlers but agree it's not a likely scenario and would probably be easy to find if it collided. and we can just leave a note, i typically use //@GOTCHA or //@BUG or //@REFACTOR to mark these types of things. a quick workaround might be to check if it has a value, and console.warn that it is being overwritten? but i'm happy with it as-is, just wanted to bring up the observation and I know i've had uis where there are multiple things being edited at once

Reply:
Ah I just merged this, but that's a great call about the console.warn - i'll add that into the following PR here https://github.com/opticdev/optic/pull/682